keps/sig-api-machinery/2855-automated-stored-versions/README.md

```diff
@@ -0,0 +1,567 @@
+
+## Motivation
+
+It is desirable for the [kube-storage-version-migrator] to modify a CRD's `status.storedVersions` array after a successful migration rather than relying on a cluster admin to manually perform the same steps.
+
+### Goals
+
+- Allow the [kube-storage-version-migrator] to update a CRD's `status.storedVersions`  with the list of stored versions after a successful migration.
+- Ensure that the kube-storage-version-migrator has successfully done its job before removing a stored version from the status of the CRD.
+
+### Non-Goals
+
+- Modify a CRD's `status.storedVersions` field if the migration is not successful.
+
+## Proposal
+
+In order to safely modify a CRD's `status.storedVersions` field, the [kube-storage-version-migrator] would need to prevent changes to the CRD during a migration. CRD's could be marked as migrating using an annotation and changes to the CRD could be prevented by means of an [Validating Admission Webhook](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/).
+
+The validating admission webhook would need to:
+
+- Intercept all Update and Patch operations against CRDs that include the `ksvm.k8s.io/migrating=true` annotation.
+- Allow `/status` requests from the [kube-storage-version-migrator] service account.
+- Allow setting or removing the migrating annotation by the [kube-storage-version-migrator] service account.
+- Prevent other changes to the CRD.
+
+A summary of the new migration workflow can be found below:
+
+1. The [kube-storage-version-migrator] notices a new stored version for the `foo` CRD.
+2. The  `ksvm.k8s.io/migrating=true`  annotation is applied to the`foo` CRD that defines the resource being migrated.
+3. The `StorageVersionMigration` CR is created for the resource introduced by the `foo` CRD.
+4. The [kube-storage-version-migrator] begins migrating all `foo` resources as it does today.
+5. When a migration is successful, the `foo` CRD's `status.storedVersion` is patched to only include the new stored version.
+6. The `ksvm.k8s.io/migrating=true` annotation is removed from the `foo` CRD.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster admin, I would like the [kube-storage-version-migrator] to automatically update a CRD's `status.storeVersions` after a successful a migration.
+
+### Notes/Constraints/Caveats (Optional)
+
+- The introduction of an admission webhook would require a valid CA, possibly managed by [cert-manager](https://github.com/jetstack/cert-manager). At a minimum, documentation should be provided that outlines how to provide a cert for the webhook.
+- The [kube-storage-version-migrator] [currently](https://github.com/kubernetes-sigs/kube-storage-version-migrator/blob/acdee30ced218b79e39c6a701985e8cd8bd33824/manifests/namespace-rbac.yaml) uses the `default` service account, a unique service account should be introduced so the webhook may reject requests not originating from the [kube-storage-version-migrator].
```

Comment:
With https://github.com/kubernetes/enhancements/pull/2856/files#r687573313, we can get rid of the service account requirement. Let everybody do any kind of changes, but block storage version changes for every actor (including ksvm during a migration.

Reply:
I've removed the mention of a new service account, thanks for the design suggestion!